src/Features/LanguageServer/Protocol/Handler/RequestContext.cs

```diff
@@ -25,9 +34,25 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler
         public string? ClientName { get; }
 
         public RequestContext(ClientCapabilities clientCapabilities, string? clientName)
+            : this(null, null, clientCapabilities, clientName)
         {
+        }
+
+        internal RequestContext(Solution? solution, Action<Solution>? solutionUpdater, ClientCapabilities clientCapabilities, string? clientName)
+        {
+            Solution = solution;
+            _solutionUpdater = solutionUpdater;
             ClientCapabilities = clientCapabilities;
             ClientName = clientName;
         }
+
+        /// <summary>
+        /// Allows a mutating request to provide a new solution snapshot that all subsequent requests should use.
+        /// </summary>
+        public void UpdateSolution(Solution solution)
+        {
+            Contract.ThrowIfNull(_solutionUpdater, "Mutating solution not allowed in a non-mutating request handler");
+            _solutionUpdater?.Invoke(solution);
```

Comment:
```suggestion
            _solutionUpdater.Invoke(solution);
```

Reply:
also, this is presumably a one-and-done sort of thing.  it might be good to store if we've applied the func, and throw if called twice.

Reply:
I thought about this, and I'm not sure I see the benefit of throwing over just taking the last updated solution that we're given. I'm going to leave this decision for later, after we actually have a handler that mutates, and see if which form fits better.